commit 6761549c5a719f541a850f243318a1212b76c0a6
Author: Sridhar Seshasayee <sridhar.seshasayee@ibm.com>
Date:   Wed May 6 20:41:33 2026 +0530

    mgr/DaemonServer: Aggregate and globally sort OSDs for ok-to-upgrade
    
    The 'ok-to-upgrade' command output sorting did not scale accurately
    when target CRUSH buckets contained multiple child buckets (e.g., a
    chassis containing multiple hosts). OSDs were previously sorted
    individually per child bucket and appended sequentially. This created
    fragmented, per-host sort segments rather than a globally sorted list
    for the parent bucket.
    
    Changes:
    
    1. Fix the issue above by aggregating all child OSDs into a single vector prior
    to executing a single, global sort operation based on PG counts. Additionally,
    optimize memory efficiency and future-proof the logic by reserving continuous
    vector blocks to avoid dynamic heap reallocations.
    
    2. Add integration tests with chassis and rack based CRUSH hierarchies which
    verifies the ok-to-upgrade functionality. In addition, the tests crucially
    verify the order of OSDs returned is according to the ascending order of
    acting PG count. Additionally, make minor fix-ups to lines that determine the
    length of a list in JSON response by removing the redundant "| bc".
    
    Fixes: https://tracker.ceph.com/issues/77272
    Signed-off-by: Sridhar Seshasayee <sridhar.seshasayee@ibm.com>

diff --git a/src/mgr/DaemonServer.cc b/src/mgr/DaemonServer.cc
index afe0a0c37fa..a7ddac0868d 100644
--- a/src/mgr/DaemonServer.cc
+++ b/src/mgr/DaemonServer.cc
@@ -1276,6 +1276,7 @@ int DaemonServer::_populate_crush_bucket_osds(
   } else if (bucket_type_str == "host" || bucket_type_str == "osd") {
     bucket_names.push_back(item_name);
   }
+
   // The following struct is to help re-order the
   // osds based on the number of pgs on them.
   struct pgs_per_osd {
@@ -1283,10 +1284,8 @@ int DaemonServer::_populate_crush_bucket_osds(
     size_t num_pgs;
   };
   std::vector<pgs_per_osd> child_bucket_pgs_per_osd;
-  // get osds under each child bucket
+  // get osds under each child bucket and associate with their PG counts
   for (const auto &name : bucket_names) {
-    // Clear the items for the current child bucket
-    child_bucket_pgs_per_osd.clear();
     std::set<int> tmp_bucket_osds;
     r = osdmap.get_osds_by_bucket_name(name, &tmp_bucket_osds);
     if (r < 0) {
@@ -1300,39 +1299,27 @@ int DaemonServer::_populate_crush_bucket_osds(
       dout(20) << os.str() << dendl;
       return r;
     }
-
-    // Special case when bucket contains only 1 osd
-    if (tmp_bucket_osds.size() == 1) {
-      for (const auto &osd : tmp_bucket_osds) {
-        crush_bucket_osds.push_back(osd);
-      }
-      dout(20) << "picked osd: " << tmp_bucket_osds
-               << " from bucket: " << name << dendl;
-      continue;
-    }
-    /**
-     * The osds in this bucket are further re-ordered based on the
-     * number of pgs (ascending) they host. This helps optimize
-     * the result of _check_offlines_pgs() down the line.
-     */
     for (const auto &osd : tmp_bucket_osds) {
       child_bucket_pgs_per_osd.push_back({osd, pgmap.get_num_pg_by_osd(osd)});
     }
-    // Sort once after all data is added
-    std::sort(child_bucket_pgs_per_osd.begin(), child_bucket_pgs_per_osd.end(),
-              [](const pgs_per_osd& a, const pgs_per_osd& b) {
-        return std::tie(a.num_pgs, a.osd_id) < std::tie(b.num_pgs, b.osd_id);
-    });
-    /**
-     * The sorted osds are finally pushed to the passed crush_bucket_osds
-     * vector where osds are maintained according to the child order.
-     */
-    for (const auto &item : child_bucket_pgs_per_osd) {
-      crush_bucket_osds.push_back(item.osd_id);
-    }
     dout(20) << "picked osds: " << tmp_bucket_osds
              << " from bucket: " << name << dendl;
   }
+
+  /**
+   * Sort all collected osds globally based on the number of pgs (ascending)
+   * they host and update the crush_bucket_osds vector with the same order.
+   */
+  std::sort(child_bucket_pgs_per_osd.begin(), child_bucket_pgs_per_osd.end(),
+            [](const pgs_per_osd& a, const pgs_per_osd& b) {
+      return std::tie(a.num_pgs, a.osd_id) < std::tie(b.num_pgs, b.osd_id);
+  });
+  crush_bucket_osds.reserve(
+    crush_bucket_osds.size() + child_bucket_pgs_per_osd.size());
+  for (const auto &item : child_bucket_pgs_per_osd) {
+    crush_bucket_osds.push_back(item.osd_id);
+  }
+
   return r;
 }
 
